presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/dereference/PushDownDereferencesUtil.java

```diff
@@ -0,0 +1,177 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner.iterative.rule.dereference;
+
+import com.google.common.collect.ImmutableList;
+import io.prestosql.spi.type.Type;
+import io.prestosql.sql.planner.PlanNodeIdAllocator;
+import io.prestosql.sql.planner.Symbol;
+import io.prestosql.sql.planner.TypeAnalyzer;
+import io.prestosql.sql.planner.iterative.Rule;
+import io.prestosql.sql.planner.plan.Assignments;
+import io.prestosql.sql.planner.plan.PlanNode;
+import io.prestosql.sql.planner.plan.ProjectNode;
+import io.prestosql.sql.tree.DefaultExpressionTraversalVisitor;
+import io.prestosql.sql.tree.DereferenceExpression;
+import io.prestosql.sql.tree.Expression;
+import io.prestosql.sql.tree.LambdaExpression;
+import io.prestosql.sql.tree.SymbolReference;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.prestosql.sql.planner.SymbolsExtractor.extractAll;
+
+/**
+ * Provides helper methods to push down dereferences in the query plan.
+ */
+public class PushDownDereferencesUtil
+{
+    private PushDownDereferencesUtil() {}
+
+    /**
+     * Create new symbols for those dereference expressions in the projections of {@param projectNode}, whose base symbol comes
+     * from output symbols of the sources of {@param projectNode}'s child.
+     */
+    static Map<DereferenceExpression, Symbol> validPushdownThroughProject(
+            Rule.Context context,
+            ProjectNode projectNode,
+            PlanNode child,
+            TypeAnalyzer typeAnalyzer)
+    {
+        Map<DereferenceExpression, Symbol> allDereferencesInProject = validDereferences(projectNode.getAssignments().getExpressions(), context, typeAnalyzer, true);
+
+        Set<Symbol> childSourceSymbols = child.getSources().stream()
+                .flatMap(node -> node.getOutputSymbols().stream())
+                .collect(toImmutableSet());
+
+        return allDereferencesInProject.entrySet().stream()
+                .filter(entry -> childSourceSymbols.contains(getBase(entry.getKey())))
+                .collect(toImmutableMap(Map.Entry::getKey, Map.Entry::getValue));
```

Comment:
We can't do this generically. The meaning of how the outputs of a node map to its inputs is node-specific, and we can't make inferences based on the fact that symbols are named the same way. It is legal for symbols to be named the same as long as they have the same type, even if they don't represent the same concept.

Also, for nodes such as Union, there's an internal remapping of inputs to outputs that changes the symbol names associated with they columns, so this would fail to identify them.

Additionally, this won't work for nodes such as `Apply` or `SemiJoin`. "Source" generally just means "an input that the operation consumes to achieve its result". In the case of Apply or Semijoin, the right side can be thought of as a subquery that gets applied for every row on the left side. In the case of Apply, there's some internal reduction operation that's done on top of the results of the subquery (semantically speaking). All this is to say that we can't generically map the dereferences in the project to the "sources" of a node -- it has to be done case-by-case.

Reply:
Thanks @martint, that makes sense. This has been incorporated now by adding separate rules for every node, rather than trying to overfit a generic pattern. I think this suggestion has also helped with code readability and reasoning.

All the rules now follow the following algorithm at a high level:  

```
ProjectNode P
    Node N
        sources S
```

* Extract dereference projections (and create new symbols) from E1 U E2 where (1) E1: projection assignment expressions and (2) E2: set of expressions being used in the node N. (eg predicate in FilterNode or function expressions in WindowNode). The logic for extraction is DereferencePushdown::validDereferences. 

* Exclude those dereference expressions, for which, the base is used as-is in the node N itself. Say the remaining dereferences are E'.   

* Pushdown E' by creating project nodes between N and S. Rewrite assignments in P and expressions in N to replace dereference expressions with new symbols.